### PR TITLE
15901 enabled Infinispan metrics

### PR DESCRIPTION
**Pull request by sschu**
Closes #15901

### CODE REVIEW

**Comment by sschu** [comment]
This should enable the Infinispan metrics. But it does not work as expected. If I make the same change on a Keycloak 20.0.1 which still has Infinispan 13 and smallrye-metrics, it works. I have not found any indication that Infinispan requires different steps to enable metrics (they just say that Micrometer Core and Micrometer Registry Prometheus must be in class path, see https://infinispan.org/docs/stable/titles/configuring/configuring.html#configuring-metrics_statistics-jmx). I will continue to play around with the config, but any help would be appreciated.
@vmuzikar @pedroigor

**Comment by pedroigor** [comment]
@sschu I'm also not sure. Perhaps smallrye-metrics/ispn has something else that automatically includes ISPN metrics into the registry.

I need to look at it more to see what we can do too.

But I think we don't need to provide a specific file to only enable statistics to the caches. We can do it programmatically here https://github.com/keycloak/keycloak/blob/605b51890ec32908594c045c91f08717220c46e8/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java#L74.

**Comment by sschu** [comment]
@pedroigor After some pretty extensive debugging sessions I realised that Infinispan ist just using its own MeterRegistries instead of the one provided by Quarkus.

Before the switch to MicroMeter, they retrieved the global registry like that: https://github.com/infinispan/infinispan/blob/1eedadc215d83e60ac2417f4dcea48c9a236a6e7/core/src/main/java/org/infinispan/metrics/impl/MetricsCollectorFactory.java#L42

After the switch (https://github.com/infinispan/infinispan/commit/0399294a33), they are creating their own registries and fill them with metrics.

However, I found https://issues.redhat.com/browse/ISPN-14341 and even a backport to Infinispan 14: https://github.com/infinispan/infinispan/pull/10483. So I assume this should be fixed with https://github.com/keycloak/keycloak/pull/16066.

Once this is merged I will have another look and rewrite the cache configuration logic to not need additional XML config files.

**Comment by pedroigor** [comment]
@sschu Thanks! Sounds good to me.

**Comment by vmuzikar** [comment]
CC @keycloak/storage-x

**Comment by ahus1** [comment]
@sschu - I can confirm that the metrics are not available. 

I looked around and found a `RootMeterRegistryBuildItem` which should give access to the meter registry. Trying that, it seems that I miss some essential Quarkus knowledge as after adding it to `configureInfinispan()` the build fails with 

> No producers for required item class io.quarkus.micrometer.deployment.RootMeterRegistryBuildItem

What did I miss?

**Comment by sschu** [comment]
@ahus1 I hope you mean "now available"? ;)

**Comment by ahus1** [comment]
@sschu - yes, sorry, I stand corrected.

**Comment by sschu** [comment]
I also don't really know how the Quarkus build process works. I can see that while debugging, some proxy object is actually set in this `RootMeterRegistryBuildItem` but it is apparently never retrieved. I was also more concerned about the way the registry is passed to Infinispan, that looked somewhat strange to me.

**Comment by sschu** [comment]
@ahus1 after quite some reading and debugging, I found a likely explanation for why the `RootMeterRegistryBuildItem` does not work. Theoretically, it should work just by passing it at a parameter to `configureInfinispan()`. When I do that, I also get the error `No producers for required item class io.quarkus.micrometer.deployment.RootMeterRegistryBuildItem`. When I run the build with tests, I get a longer trace hinting a cyclic dependency of the build steps.

Looking a bit further, I saw this remark: https://github.com/quarkusio/quarkus/blob/e58e483cd298818ceddd7f5fdad2e55879e58ec8/extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerProcessor.java#L167-L173

This essentially means that that the `RootMeterRegistryBuildItem` is only created when the CDI-container if fully initialised since it takes a `BeanContainerBuildItem` as parameter. At the same time, `configureInfinispan()` produces `SyntheticBeanBuildItem` which I assume means this has to happen before the CDI-container is fully initialised.

So I would for now go with just accessing the global registry directly. Or do you have an idea for an alternative solution?

**Comment by ahus1** [comment]
@sschu - thanks for pointing me at this code. I now understand why the code you added to access the meter registry works. So I'm happy to approve this PR. I wish I would have a better solution, but I don't and it might be due to knowledge of Quarkus in these parts. 

I see that this change adds ~796 metrics when running with the standard H2 + Infinispan setup. So I wonder if people might want to disable cache metrics via an option. Still this would be a follow-up PR, and people can filter out those metrics via a rewrite rule.

I also see that the name of the cache appears twice - once in the name of the metrics, and once as a dimension. Also having the node name in there is strange.

> vendor_cache_manager_keycloak_cache_sessions_statistics_hits{cache="sessions",node="dockingstation-2738",} 16.0

Again, this is nothing this PR can fix - this should be fixed upstream, and people might use a rewrite rule in the meantime.

@pedroigor - I suppose this PR is good enough and ready to be merged. Please let me know your thoughts. Thanks!

**Review by ahus1 — DISMISSED** [review]
Thanks - finally some metrics for Infinispan!

**Comment by sschu** [comment]
@ahus1 Thanks! I am somewhat confident there is not a better way to access the global registry. And although the `builder`-code to pass the registry to Infinispan does not look very natural to me, I assume it's the way to go. It is definitely an Infinispan-only thing and not Quarkus magic I am missing because Infinispan is not based on Quarkus.

I was wondering if I should add a test to check for existence of at least one metric so we notice if something does not work anymore with this not so usual configuration style.

Can you also add a Milestone KC21 label as not adding this to 21 means no more Infinispan metrics (KC20 still had small rye Infinispan metrics).

**Comment by ahus1** [comment]
@sschu - milestone added. 

Regarding the tests: They are always welcome, at the same time I don't put too much emphasis on this as this is only metrics which are at best incubating at the moment (no Keycloak metrics, only 3rd-party metrics). Feel free to add a test if you'd like to see one.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java [hunk 1]
```diff
@@ -86,6 +97,10 @@ private DefaultCacheManager startCacheManager() {
         return new DefaultCacheManager(builder, isStartEagerly());
     }
 
+    private boolean isMetricsEnabled() {
```
Comment (pedroigor) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java:100]: We can avoid looking up if a build option is set by resolving it at build time. For that, you can change this line to create the `CacheManagerFactory` instance from the recorded by passing the value for the `metrics-enabled` option.

Sounds like a nitpick but we need to try to reduce any unnecessary steps during bootstrapping and favor building the configuration during re-augmentation / build time.

Looking up options at runtime is a bit expensive as it ends up going through all MP config sources, etc.

**Review by pedroigor — COMMENTED** [review]
@ahus1 @sschu LGTM too. Thanks for this one.

Just a comment about how we can avoid executing unnecessary steps at runtime when checking if `metrics-enabled` is set and a test assertion.

Please, see https://github.com/pedroigor/keycloak/tree/tmp-bosh.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java (hunk 1 above)
Comment (sschu) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java]: Thanks for your suggestions! I applied them (and learned something).

**Review by pedroigor — APPROVED** [review]
Thanks!

**Comment by sschu** [comment]
@ahus1 I can take this. I am not sure I will be able to contribute at Infinispan but I can at least start the discussion.

**Pull request merged**